### PR TITLE
Added overlay renderered for GC, snapshot and memory reset events.

### PR DESCRIPTION
**Pull request by terrylucas**
- Fixed GC happening when getAllocationProfile is called.
- Fixed Legend computation of range to match adt-ui (bug but we use the same workaround).
- Added overlay for rendering of events: GC, snapshot and memory reset in timeline.
- Added glyphs for the 3 events and added explicit manual GC.

### CODE REVIEW

File: flutter-studio/src/io/flutter/profiler/FlutterAllMemoryData.java [hunk 1]
```diff
@@ -43,7 +49,11 @@
     public ThreadSafeData() { }
 
     @Override
-    public List<SeriesData<Long>> getDataForXRange(Range xRange) {
+    public List<SeriesData<Long>> getDataForXRange(Range range) {
+      // NOTE: adt-ui is off on the range, need to adjust. The view is beyojd the last item in the series
```
Comment (jacob314) [inline flutter-studio/src/io/flutter/profiler/FlutterAllMemoryData.java]: typo in this comment and perhaps clarify the comment a bit.
beyojd --> beyond

File: flutter-studio/src/io/flutter/profiler/FlutterAllMemoryData.java [hunk 2]
```diff
@@ -109,6 +119,15 @@ public void handleIsolatesInfo(VM vm, List<HeapMonitor.IsolateObject> isolates)
       public void handleGCEvent(IsolateRef iIsolateRef,
                                 HeapMonitor.HeapSpace newHeapSpace,
                                 HeapMonitor.HeapSpace oldHeapSpace) {
+        // TODO(terry): This event isn't being called?
```
Comment (jacob314) [inline flutter-studio/src/io/flutter/profiler/FlutterAllMemoryData.java]: is this comment obsolete? Looks like GC events are showing up.

File: flutter-studio/src/io/flutter/profiler/FlutterStudioMonitorStageView.java [hunk 3]
```diff
@@ -584,11 +587,38 @@ public void received(AllocationProfile response) {
...
+
+    vmService.getAllocationProfile(isolateId, "full", null, new AllocationProfileConsumer() {
+      @Override
+      public void onError(RPCError error) {
+        LOG.error("Allocation Profile during gcNow - " + error.getDetails());
+        future.completeExceptionally(new RuntimeException(error.toString()));
+      }
+      @Override
+      public void received(AllocationProfile response) {
+        // TODO(terry): Add GC Icon to overlay - should happen in FlutterAllMemoryData however handleGCEvent is never called.
```
Comment (jacob314) [inline flutter-studio/src/io/flutter/profiler/FlutterStudioMonitorStageView.java]: aren't these icons showing up?

File: flutter-studio/src/io/flutter/profiler/FlutterStudioMonitorStageView.java [hunk 4]
```diff
@@ -584,11 +587,38 @@ public void received(AllocationProfile response) {
...
+        view.getStage().recordGC();
+      }
+    });
+  }
+
+  // TODO(terry): Remove for testing only.
```
Comment (jacob314) [inline flutter-studio/src/io/flutter/profiler/FlutterStudioMonitorStageView.java]: remove

File: flutter-studio/src/io/flutter/profiler/FlutterStudioMonitorStageView.java [hunk 5]
```diff
@@ -657,6 +687,11 @@ public void selectionCleared() {
       rssDataSeries = stage.getRSSDataSeries();
     }
 
+    FlutterAllMemoryData.ThreadSafeData gcDataSeries = stage.getGcDataSeries();
```
Comment (jacob314) [inline flutter-studio/src/io/flutter/profiler/FlutterStudioMonitorStageView.java]: are these 3 variables needed. I don't see them being used but perhaps I am missing something.

Comment (jacob314) [reply flutter-studio/src/io/flutter/profiler/FlutterStudioMonitorStageView.java]: nm. perhaps move these variables down closer to where they are used.

File: flutter-studio/src/io/flutter/profiler/FlutterAllMemoryData.java (hunk 2 above)
Comment (terrylucas) [reply flutter-studio/src/io/flutter/profiler/FlutterAllMemoryData.java]: They're showing up but as a result of the explicit manual GC being pushed the handleGCEvent is not being fired.  So I'd prefer to handle the event here once this works (VMService issue).

File: flutter-studio/src/io/flutter/profiler/FlutterStudioMonitorStageView.java (hunk 3 above)
Comment (terrylucas) [reply flutter-studio/src/io/flutter/profiler/FlutterStudioMonitorStageView.java]: Ditto on the handleGCEvent.

File: flutter-studio/src/io/flutter/profiler/FlutterStudioMonitorStageView.java [hunk 6]
```diff
@@ -703,6 +738,61 @@ public void selectionCleared() {
...
+
+    mLineChart.addCustomRenderer(resetRenderer);
+    overlayComponent.addDurationDataRenderer(resetRenderer);
+
+    // Build the Snapshot icon rendering.
+    DurationDataModel<SnapshotData> snapshotModel = new DurationDataModel<>(new RangedSeries<>(viewRange, new SnapshotDataSeries(snapshotDataSeries)));
+    snapshotModel.setAttachedSeries(maxMemoryRange, Interpolatable.SegmentInterpolator);
+    snapshotModel.update(0);
+
+    DurationDataRenderer<SnapshotData> snapshotRenderer = new DurationDataRenderer.Builder<>(snapshotModel, Color.BLACK)
```
Comment (jacob314) [inline flutter-studio/src/io/flutter/profiler/FlutterStudioMonitorStageView.java]: make sure you are using JBColor everywhere. Also test what this UI looks like in AndroidStudio dark mode. Without JBColor, the UI will likely be hard to read.

File: flutter-studio/src/io/flutter/profiler/FlutterStudioMonitorStageView.java [hunk 7]
```diff
@@ -703,6 +738,61 @@ public void selectionCleared() {
...
+      .setIcon(FlutterIcons.Snapshot)
+      .setHostInsets(new Insets(Y_AXIS_TOP_MARGIN, 0, 0, 0))
+      .build();
+
+    mLineChart.addCustomRenderer(snapshotRenderer);
+    overlayComponent.addDurationDataRenderer(snapshotRenderer);
+
+    final JPanel overlayPanel = new JBPanel(new BorderLayout());
+    overlayPanel.setOpaque(false);
+    overlayPanel.setBorder(BorderFactory.createEmptyBorder(0, 0, 0, 0));
```
Comment (jacob314) [inline flutter-studio/src/io/flutter/profiler/FlutterStudioMonitorStageView.java]: nit: is it necessary to set an empty border?

File: flutter-studio/src/io/flutter/profiler/FlutterStudioMonitorStageView.java (hunk 5 above)
Comment (terrylucas) [reply flutter-studio/src/io/flutter/profiler/FlutterStudioMonitorStageView.java]: Yes when the DurationModel is created e.g.,

      new DurationDataModel<>(new RangedSeries<>(viewRange, new ResetStatsDataSeries(resetDataSeries)));

etc.

File: flutter-studio/src/io/flutter/profiler/GcDurationData.java [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+import com.android.tools.adtui.model.DefaultDurationData;
+
+public final class GcDurationData extends DefaultDurationData {
+   public GcDurationData(long duration) {
+      super(duration);
+   }
+
+   @Override
+   public String toString() {
+     return "GC Event";
```
Comment (jacob314) [inline flutter-studio/src/io/flutter/profiler/GcDurationData.java]: This toString seems odd. Perhaps remove or include the duration in the toString.

File: flutter-studio/src/io/flutter/profiler/GcStatsDataSeries.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+import com.android.tools.adtui.model.DataSeries;
+import com.android.tools.adtui.model.Range;
+import com.android.tools.adtui.model.SeriesData;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import org.jetbrains.annotations.NotNull;
+
+public final class GcStatsDataSeries implements DataSeries<GcDurationData> {
+  FlutterAllMemoryData.ThreadSafeData gcData;
```
Comment (jacob314) [inline flutter-studio/src/io/flutter/profiler/GcStatsDataSeries.java]: Can you make this final?

File: flutter-studio/src/io/flutter/profiler/ResetStatsDataSeries.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+  public ResetStatsDataSeries(FlutterAllMemoryData.ThreadSafeData gcDataSeries) {
+    resetData = gcDataSeries;
+  }
+
+  @Override
+  public List<SeriesData<ResetData>> getDataForXRange(@NotNull Range xRange) {
+    List<SeriesData<ResetData>> seriesData = new ArrayList<>();
+
+    List<SeriesData<Long>> rawGcValues = resetData.getDataForXRange(xRange);
+    for (SeriesData<Long>  sample : rawGcValues) {
```
Comment (jacob314) [inline flutter-studio/src/io/flutter/profiler/ResetStatsDataSeries.java]: make sure the code formatter has been run on all code. This indentation seems a bit non-standard.

File: flutter-studio/src/io/flutter/profiler/SnapshotData.java [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+import com.android.tools.adtui.model.DefaultDurationData;
+
+public final class SnapshotData extends DefaultDurationData {
+  public SnapshotData(long duration) {
+    super(duration);
+  }
+
+  @Override
+  public String toString() {
+    return "Reset Event";
```
Comment (jacob314) [inline flutter-studio/src/io/flutter/profiler/SnapshotData.java]: This toString seems like a bug. I saw another toString that was reset event. Perhaps remove these toStrings unless they impact the UI.

File: flutter-studio/src/io/flutter/profiler/FlutterAllMemoryData.java (hunk 1 above)
Comment (terrylucas) [reply flutter-studio/src/io/flutter/profiler/FlutterAllMemoryData.java]: Done.

File: flutter-studio/src/io/flutter/profiler/FlutterStudioMonitorStageView.java (hunk 4 above)
Comment (terrylucas) [reply flutter-studio/src/io/flutter/profiler/FlutterStudioMonitorStageView.java]: Comment was not accurate should read:
  // TODO(terry): Need to remove when we 3.2 of AS isn't supported only post 3.2 have this png.

File: flutter-studio/src/io/flutter/profiler/FlutterStudioMonitorStageView.java (hunk 6 above)
Comment (terrylucas) [reply flutter-studio/src/io/flutter/profiler/FlutterStudioMonitorStageView.java]: Done.

File: flutter-studio/src/io/flutter/profiler/GcStatsDataSeries.java (hunk 9 above)
Comment (terrylucas) [reply flutter-studio/src/io/flutter/profiler/GcStatsDataSeries.java]: No one else is subclassing these type of classes - right?

File: flutter-studio/src/io/flutter/profiler/GcDurationData.java (hunk 8 above)
Comment (terrylucas) [reply flutter-studio/src/io/flutter/profiler/GcDurationData.java]: Gone.

File: flutter-studio/src/io/flutter/profiler/FlutterStudioMonitorStageView.java (hunk 7 above)
Comment (terrylucas) [reply flutter-studio/src/io/flutter/profiler/FlutterStudioMonitorStageView.java]: I don't think so - removed.

**Comment by terrylucas** [comment]
Thanks, PTAL.

File: flutter-studio/src/io/flutter/profiler/ResetStatsDataSeries.java (hunk 10 above)
Comment (terrylucas) [reply flutter-studio/src/io/flutter/profiler/ResetStatsDataSeries.java]: re-ran formatter.

File: flutter-studio/src/io/flutter/profiler/SnapshotData.java (hunk 11 above)
Comment (terrylucas) [reply flutter-studio/src/io/flutter/profiler/SnapshotData.java]: Diito on all these toStrings removed.

**Pull request merged**